File: admin-portal/src/services/cachedApiService.ts
import apiService from './apiService';
import { 
  shortCache, 
  mediumCache, 
  longCache, 
  staticCache, 
  CACHE_KEYS 
} from './cacheService';

// Cache strategies for different endpoints
const CACHE_STRATEGIES = {
  // Auth - short cache for security
  '/api/auth/validate': { cache: shortCache, ttl: 30 * 1000 },
  '/api/auth/login': { cache: null }, // No cache for login
  
  // Companies - medium cache
  '/api/super-admin/companies': { cache: mediumCache, ttl: 2 * 60 * 1000 },
  '/api/companies': { cache: mediumCache, ttl: 2 * 60 * 1000 },
  
  // Subscription plans - long cache (rarely change)
  '/api/super-admin/subscription-plans': { cache: longCache, ttl: 15 * 60 * 1000 },
  
  // Dashboard stats - short cache (frequently updated)
  '/api/super-admin/dashboard/stats': { cache: shortCache, ttl: 60 * 1000 },
  
  // Analytics - medium cache
  '/api/super-admin/analytics': { cache: mediumCache, ttl: 5 * 60 * 1000 },
  
  // Users - medium cache
  '/api/super-admin/users': { cache: mediumCache, ttl: 2 * 60 * 1000 },
  
  // Settings - long cache
  '/api/super-admin/settings': { cache: longCache, ttl: 10 * 60 * 1000 },
  '/api/admin/settings': { cache: longCache, ttl: 10 * 60 * 1000 },
  
  // Notifications - short cache
  '/api/super-admin/notifications': { cache: shortCache, ttl: 30 * 1000 },
  
  // Monitoring - short cache
  '/api/monitoring/health': { cache: shortCache, ttl: 30 * 1000 },
  '/api/monitoring/errors': { cache: shortCache, ttl: 30 * 1000 },
  '/api/monitoring/performance': { cache: shortCache, ttl: 30 * 1000 }
};

// Helper to get cache strategy for URL
function getCacheStrategy(url: string) {
  // Find matching strategy
  for (const [pattern, strategy] of Object.entries(CACHE_STRATEGIES)) {
    if (url.includes(pattern)) {
      return strategy;
    }
  }
  
  // Default to no cache
  return { cache: null, ttl: 0 };
}

// Helper to generate cache key
function generateCacheKey(method: string, url: string, data?: any): string {
  const baseKey = `${method}:${url}`;
  
  // Include data in key for POST/PUT/PATCH requests
  if (data && ['POST', 'PUT', 'PATCH'].includes(method.toUpperCase())) {
    const dataHash = JSON.stringify(data).slice(0, 100); // Limit hash length
    return `${baseKey}:${dataHash}`;
  }
  
  return baseKey;
}

// Cached API service
export const cachedApiService = {
  // GET with caching
  async get<T>(url: string, config?: any): Promise<T> {
    const strategy = getCacheStrategy(url);
    const cacheKey = generateCacheKey('GET', url);
    
    // Check cache first
    if (strategy.cache) {
      const cached = strategy.cache.get<T>(cacheKey);
      if (cached !== null) {
        console.log(`Cache HIT: ${url}`);
        return cached;
      }
    }
    
    // Fetch from API
    console.log(`Cache MISS: ${url}`);
    const data = await apiService.get<T>(url, config);
    
    // Store in cache
    if (strategy.cache) {
      strategy.cache.set(cacheKey, data, strategy.ttl);
    }
    
    return data;
  },

  // POST without caching (mutations)
  async post<T>(url: string, data?: any, config?: any): Promise<T> {
    const result = await apiService.post<T>(url, data, config);
    
    // Invalidate related caches after mutations
    this.invalidateRelatedCaches(url, data);
    
    return result;
  },

  // PUT without caching (mutations)
  async put<T>(url: string, data?: any, config?: any): Promise<T> {
    const result = await apiService.put<T>(url, data, config);
    
    // Invalidate related caches after mutations
    this.invalidateRelatedCaches(url, data);
    
    return result;
  },

  // PATCH without caching (mutations)
  async patch<T>(url: string, data?: any, config?: any): Promise<T> {
    const result = await apiService.patch<T>(url, data, config);
    
    // Invalidate related caches after mutations
    this.invalidateRelatedCaches(url, data);
    
    return result;
  },

  // DELETE without caching (mutations)
  async delete<T>(url: string, config?: any): Promise<T> {
    const result = await apiService.delete<T>(url, config);
    
    // Invalidate related caches after mutations
    this.invalidateRelatedCaches(url);
    
    return result;
  },

  // Upload without caching
  async upload<T>(url: string, formData: FormData, config?: any): Promise<T> {
    const result = await apiService.upload<T>(url, formData, config);
    
    // Invalidate related caches after mutations
    this.invalidateRelatedCaches(url);
    
    return result;
  },

  // Cache invalidation methods
  invalidateRelatedCaches(url: string, data?: any) {
    // Invalidate based on URL patterns
    if (url.includes('/companies')) {
      mediumCache.delete(CACHE_KEYS.COMPANIES_LIST);
      console.log('Invalidated companies cache');
    }
    
    if (url.includes('/users')) {
      mediumCache.delete(CACHE_KEYS.USERS_LIST);
      console.log('Invalidated users cache');
    }
    
    if (url.includes('/settings')) {
      longCache.delete(CACHE_KEYS.SYSTEM_SETTINGS);
      console.log('Invalidated settings cache');
    }
    
    if (url.includes('/analytics')) {
      mediumCache.delete(CACHE_KEYS.ANALYTICS_DATA('30d'));
      console.log('Invalidated analytics cache');
    }
    
    if (url.includes('/dashboard')) {
      shortCache.delete(CACHE_KEYS.DASHBOARD_STATS);
      console.log('Invalidated dashboard cache');
    }
  },

  // Manual cache invalidation
  invalidateCache(pattern: string) {
    [shortCache, mediumCache, longCache, staticCache].forEach(cache => {
      // This is a simplified invalidation - in production you'd want more sophisticated pattern matching
      console.log(`Invalidating cache pattern: ${pattern}`);
    });
  },

  // Clear all caches
  clearAllCaches() {
    shortCache.clear();
    mediumCache.clear();
    longCache.clear();
    staticCache.clear();
    console.log('All caches cleared');
  },

  // Get cache statistics
  getCacheStats() {
    return {
      short: shortCache.getStats(),
      medium: mediumCache.getStats(),
      long: longCache.getStats(),
      static: staticCache.getStats()
    };
  },

  // Preload important data
  async preloadData() {
    try {
      console.log('Preloading important data...');
      
      // Preload subscription plans (long cache)
      await this.get('/api/super-admin/subscription-plans');
      
      // Preload system settings (long cache)
      await this.get('/api/super-admin/settings');
      
      console.log('Data preloading completed');
    } catch (error) {
      console.warn('Data preloading failed:', error);
    }
  }
};

export default cachedApiService; 

File: sns_rooster/lib/services/cached_api_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'api_service.dart';
import 'cache_service.dart';
import '../config/api_config.dart';

/// Cached API Service for Flutter App
/// Wraps the existing API service with intelligent caching

class CachedApiService {
  final ApiService _apiService;
  final CacheService _cacheService;

  CachedApiService({
    ApiService? apiService,
    CacheService? cacheService,
  })  : _apiService = apiService ?? ApiService(baseUrl: ApiConfig.baseUrl),
        _cacheService = cacheService ?? CacheService();

  // Cache strategies for different endpoints
  static const Map<String, Duration> _cacheStrategies = {
    // Auth - short cache for security
    '/auth/validate': CacheService.shortTTL,
    '/auth/login': Duration.zero, // No cache for login

    // Company - medium cache
    '/companies': CacheService.mediumTTL,
    '/companies/features': CacheService.longTTL,

    // Attendance - short cache (frequently updated)
    '/attendance/today': CacheService.shortTTL,
    '/attendance/summary': CacheService.shortTTL,
    '/attendance/records': CacheService.mediumTTL,

    // Leave - medium cache
    '/leave/info': CacheService.mediumTTL,
    '/leave/types': CacheService.longTTL,
    '/leave/records': CacheService.mediumTTL,

    // Payroll - medium cache
    '/payroll/info': CacheService.mediumTTL,
    '/payroll/records': CacheService.mediumTTL,

    // Settings - long cache
    '/settings': CacheService.longTTL,
    '/admin/settings': CacheService.longTTL,

    // Notifications - short cache
    '/notifications': CacheService.shortTTL,

    // Analytics - medium cache
    '/analytics': CacheService.mediumTTL,
  };

  /// Get cache strategy for URL
  Duration _getCacheStrategy(String url) {
    for (final entry in _cacheStrategies.entries) {
      if (url.contains(entry.key)) {
        return entry.value;
      }
    }
    return Duration.zero; // No cache by default
  }

  /// Generate cache key
  String _generateCacheKey(String method, String url,
      [Map<String, dynamic>? data]) {
    final baseKey = '${method.toUpperCase()}:$url';

    if (data != null &&
        ['POST', 'PUT', 'PATCH'].contains(method.toUpperCase())) {
      final dataHash = jsonEncode(data).substring(0, 100); // Limit hash length
      return '$baseKey:$dataHash';
    }

    return baseKey;
  }

  /// GET with caching
  Future<ApiResponse> get(
    String url, {
    Map<String, String>? headers,
  }) async {
    final cacheStrategy = _getCacheStrategy(url);
    final cacheKey = _generateCacheKey('GET', url);

    // Check cache first
    if (cacheStrategy > Duration.zero) {
      final cached = await _cacheService.get<ApiResponse>(cacheKey);
      if (cached != null) {
        if (kDebugMode) {
          print('Cache HIT: $url');
        }
        return cached;
      }
    }

    // Fetch from API
    if (kDebugMode) {
      print('Cache MISS: $url');
    }

    final response = await _apiService.get(url);

    // Store in cache
    if (cacheStrategy > Duration.zero && response.success) {
      await _cacheService.set(cacheKey, response, cacheStrategy);
    }

    return response;
  }

  /// POST without caching (mutations)
  Future<ApiResponse> post(
    String url, {
    Map<String, dynamic>? data,
    Map<String, String>? headers,
  }) async {
    final result = await _apiService.post(url, data ?? {});

    // Invalidate related caches after mutations
    await _invalidateRelatedCaches(url, data);

    return result;
  }

  /// PUT without caching (mutations)
  Future<ApiResponse> put(
    String url, {
    Map<String, dynamic>? data,
    Map<String, String>? headers,
  }) async {
    final result = await _apiService.put(url, data);

    // Invalidate related caches after mutations
    await _invalidateRelatedCaches(url, data);

    return result;
  }

  /// PATCH without caching (mutations)
  Future<ApiResponse> patch(
    String url, {
    Map<String, dynamic>? data,
    Map<String, String>? headers,
  }) async {
    final result = await _apiService.patch(url, data ?? {});

    // Invalidate related caches after mutations
    await _invalidateRelatedCaches(url, data);

    return result;
  }

  /// DELETE without caching (mutations)
  Future<ApiResponse> delete(
    String url, {
    Map<String, String>? headers,
  }) async {
    final result = await _apiService.delete(url);

    // Invalidate related caches after mutations
    await _invalidateRelatedCaches(url);

    return result;
  }

  /// Upload without caching
  /// Note: Upload functionality not implemented in base ApiService
  /// Use direct HTTP client for file uploads
  Future<ApiResponse> upload(
    String url, {
    required Map<String, dynamic> formData,
    Map<String, String>? headers,
  }) async {
    // TODO: Implement upload functionality or use direct HTTP client
    throw UnimplementedError(
        'Upload method not implemented in base ApiService');
  }

  /// Invalidate related caches based on URL patterns
  Future<void> _invalidateRelatedCaches(String url,
      [Map<String, dynamic>? data]) async {
    // Invalidate based on URL patterns
    if (url.contains('/attendance')) {
      await _cacheService.delete(CacheKeys.todayAttendance);
      await _cacheService.delete(CacheKeys.attendanceSummary);
      if (kDebugMode) {
        print('Invalidated attendance cache');
      }
    }

    if (url.contains('/leave')) {
      await _cacheService.delete(CacheKeys.leaveInfo);
      if (kDebugMode) {
        print('Invalidated leave cache');
      }
    }

    if (url.contains('/payroll')) {
      await _cacheService.delete(CacheKeys.payrollInfo);
      if (kDebugMode) {
        print('Invalidated payroll cache');
      }
    }

    if (url.contains('/settings')) {
      await _cacheService.delete(CacheKeys.appSettings);
      await _cacheService.delete(CacheKeys.companySettings);
      if (kDebugMode) {
        print('Invalidated settings cache');
      }
    }

    if (url.contains('/companies')) {
      await _cacheService.delete(CacheKeys.companiesList);
      await _cacheService.delete(CacheKeys.companyFeatures);
      if (kDebugMode) {
        print('Invalidated companies cache');
      }
    }

    if (url.contains('/notifications')) {
      await _cacheService.delete(CacheKeys.notifications);
      if (kDebugMode) {
        print('Invalidated notifications cache');
      }
    }

    if (url.contains('/analytics')) {
      await _cacheService.invalidatePattern('analytics');
      if (kDebugMode) {
        print('Invalidated analytics cache');
      }
    }
  }

  /// Manual cache invalidation
  Future<void> invalidateCache(String pattern) async {
    await _cacheService.invalidatePattern(pattern);
    if (kDebugMode) {
      print('Manually invalidated cache pattern: $pattern');
    }
  }

  /// Clear all caches
  Future<void> clearAllCaches() async {
    await _cacheService.clear();
    if (kDebugMode) {
      print('All caches cleared');
    }
  }

  /// Get cache statistics
  Map<String, dynamic> getCacheStats() {
    return _cacheService.getStats();
  }

  /// Preload important data
  Future<void> preloadData() async {
    try {
      if (kDebugMode) {
        print('Preloading important data...');
      }

      // Preload company features
      await get('/companies/features');

      // Preload leave types
      await get('/leave/types');

      // Preload app settings
      await get('/settings');

      if (kDebugMode) {
        print('Data preloading completed');
      }
    } catch (error) {
      if (kDebugMode) {
        print('Data preloading failed: $error');
      }
    }
  }

  /// Initialize cache service
  Future<void> initialize() async {
    await _cacheService.initialize();
    await preloadData();
  }
}

// Global cached API service instance
final cachedApiService = CachedApiService();
